protocols/floodsub/src/protocol.rs

```diff
@@ -56,10 +63,28 @@ where
     type Error = FloodsubDecodeError;
     type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
 
-    fn upgrade_inbound(self, mut socket: TSocket, _: Self::Info) -> Self::Future {
+    fn upgrade_inbound(self, socket: TSocket, _: Self::Info) -> Self::Future {
         Box::pin(async move {
-            let packet = upgrade::read_length_prefixed(&mut socket, 2048).await?;
-            let rpc = rpc_proto::Rpc::decode(&packet[..]).map_err(DecodeError)?;
+            let mut framed = Framed::<TSocket, prost_codec::Codec<rpc_proto::Rpc>>::new(
+                socket,
+                prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
+            );
+            // let rpc = match framed.next().await {
+            //     Some(v) => v.unwrap(),
+            //     None => return DecodeError,
+            //     // _ => panic!(), // Better error here?
+            // };
+
+            // Better approach:
+            // let rpc = framed.next().await.unwrap_or_else(|| DecodeError);
+            // let rpc_a = framed.next().await.ok_or_else(|| DecodeError); //u nwrap_or_else(|e| (*e).clone());
+            // let rpc = rpc_a.unwrap();
+
+            // TODO: Understand what the ? is doing, and how to unwrap this without using a return (return in match causes errors in the map_err below.)
```

Comment:
`unwrap_or_else` maps `Option` to `Result` (Hint: Read its source code!) so `?` just unpacks the `Result` and returns the `Err` if present and evaluates to `Ok` otherwise, same functionality as everywhere really :)

Reply:
You may need another `?` to get rid of the 2nd `Result`!

Reply:
Ahh, thank you! I was twisted up badly on this one, but I think I've got it now.  Also, reading the docs I think I was using `unwrap_or_else` when I should have been using `ok_or_else`. (The docs make it sound like `unwrap_or_else` should be used if I have a value I can return, and `ok_or_else` when I need to returns/handle an error)

Reply:
My bad yes, it is `ok_or_else`. Sorry for sending you down that path!